File: shoggoth/src/arch/x86/src/Console.cpp
#include "Console.hpp"
#include "../../../memory.hpp"

namespace x86 {
    static Console instance;

    Console::Console(void) {
        this->setForeground(Color::LightGray);
        this->setBackground(Color::Black);
        this->xpos = 0;
        this->ypos = 0;
    }

    Console& Console::getInstance(void) {
        return instance;
    }

    void Console::setForeground(Color c) {
        this->foreground = c;
    }

    void Console::setBackground(Color c) {
        this->background = c;
    }

    void Console::putchar(char c) {
        uint16_t *ptr = (uint16_t*)0xB8000;
        uint16_t foreground = static_cast<uint8_t>(this->foreground);
        uint16_t background = static_cast<uint8_t>(this->background);

        if(c == '\n') {
            this->xpos = 0;
            this->ypos++;
            return;
        }

        if(c == '\t') {
            this->xpos += 4;
            return;
        }

        *(ptr + this->xpos + 80 * this->ypos) = ((foreground | ((background & 0x7) << 4)) << 8) | c;
        this->xpos++;

        if(this->xpos >= 80) {
            this->xpos = 0;
            this->ypos++;
            if(this->ypos >= 25) {

            }
        }
    }

    void Console::print(const char *s) {
        while(*s) {
            this->putchar(*s);
            s++;
        }
    }

    void Console::clear(void) {
        uint16_t foreground = static_cast<uint8_t>(this->foreground);
        uint16_t background = static_cast<uint8_t>(this->background);
        uint16_t value = ((foreground | ((background & 0x7) << 4)) << 8) | ' ';
        memset<uint16_t>((void*) 0x0B8000, value, 80 * 25);
        this->xpos = 0;
        this->ypos = 0;
    }
}

File: sothoth/src/arch/x86/console.rs
use console_trait;
use console_trait::Color;
use core::ptr::write_volatile;
use rlibc::memset;

pub struct Console {
    x: i8,
    y: i8,
    fg: Color,
    bg: Color
}

impl Console {
    pub fn new() -> Console {
        Console {
            x: 0,
            y: 0,
            bg: Color::Black,
            fg: Color::LightGray
        }
    }
}

impl console_trait::Console for Console {
    fn set_foreground(&mut self, fg: Color) {
        self.fg = fg;
    }

    fn set_background(&mut self, bg: Color) {
        self.bg = bg;
    }

    fn clear(&mut self) {
        let fg = self.fg as u32;
        let bg = self.bg as u32;
        let color = (fg | bg << 4) << 8;

        self.x = 0;
        self.y = 0;

        unsafe {
            memset(0xB8000 as *mut u8, ((color | 0x20) << 16) as i32, 80 * 25 * 2);
        }
    }

    fn print(&mut self, s: &str) {
        let ptr = 0xB8000 as *mut u16;
        let fg = self.fg as u16;
        let bg = self.bg as u16;
        let color = (fg | bg << 4) << 8;

        for cs in s.as_bytes() {
            let c = *cs as u8;
            let cl = *cs as u16;

            if c == b'\n' {
                self.x = 0;
                self.y += 1;
                if self.y + 1 >= 80 {
                    self.y = 0;
                }
            }
            else
            if c == b'\t' {

            }

            unsafe {
                write_volatile(ptr.offset((self.x + self.y * 80) as isize), color | cl);
            }

            self.x += 1;
        }
    }
}
